### PR TITLE
Moar TSDoc.

### PR DESCRIPTION
**Pull request by jonathan-codaio**
Getting down to ones that take a long time to write...

PTAL @ekoleda-codaio @coda/packs

### CODE REVIEW

**Comment by jonathan-codaio** [comment]
Any idea why we get these warnings with `make typedoc`?

```
Warning: NoAuthentication, defined at types.ts:134, is referenced by Authentication but not included in the documentation.
Warning: VariousAuthentication, defined at types.ts:433, is referenced by Authentication but not included in the documentation.
Warning: HeaderBearerTokenAuthentication, defined at types.ts:240, is referenced by Authentication but not included in the documentation.
Warning: CodaApiBearerTokenAuthentication, defined at types.ts:254, is referenced by Authentication but not included in the documentation.
```

I thought that typedoc pulled in the transitive closure of what was referenced in `index.ts`, but it appears to be omitting all these types that it warns about, for some reason. Some or all of those that it warns about we do want to be in the documentation.

File: helpers/url.ts [hunk 1]
```diff
@@ -2,6 +2,17 @@ import {ensureNonEmptyString} from './ensure';
...
 
+/**
+ * Helper to create a new URL by appending parameters to a base URL.
+ *
+ * The input URL may or may not having existing parameters.
+ *
+ * @example
+ * ```
+ * // Returns `"/someApi/someEndpoint?token=asdf&limit=5"`
+ * const url = withQueryParams("/someApi/someEndpoint", {token: "asdf", limit: 5});
```
Comment (ekoleda-codaio) [inline helpers/url.ts:13]: Nitpick: The style guide says to prefer `let` over `const` except when it's a file-level constant (which this URL likely isn't).

File: helpers/url.ts [hunk 2]
```diff
@@ -14,6 +25,15 @@ export function withQueryParams(url: string, params?: {[key: string]: any}): str
   return parsedUrl.toString();
 }
 
+/**
+ * Helper to take a URL string and return the parameters (if any) as a JavaScript object.
+ *
+ * @example
+ * ```
+ * // Returns `{token: "asdf", limit: "5"}`
+ * const params = getQueryParams("/someApi/someEndpoint?token=asdf&limit=5");
```
Comment (ekoleda-codaio) [inline helpers/url.ts:34]: Ditto

File: schema.ts [hunk 3]
```diff
@@ -356,29 +370,75 @@ export enum ScaleIconSet {
...
+ * which Coda should interpret as a date. Coda is able to flexibly a parse number of formal
+ * and informal string representations of dates. For maximum accuracy, consider using an
+ * ISO 8601 date string (e.g. 2021-10-29): https://en.wikipedia.org/wiki/ISO_8601.
+ */
 export interface StringDateSchema extends BaseStringSchema<ValueHintType.Date> {
+  /** Instructs Coda to render this value as a date. */
   codaType: ValueHintType.Date;
-  // A Moment date format string, such as 'MMM D, YYYY', that corresponds to a supported Coda date column format.
+  /**
+   * A Moment date format string, such as 'MMM D, YYYY', that corresponds to a supported Coda date column format.
```
Comment (ekoleda-codaio) [inline schema.ts:400]: Is it worth clarifying that this is the output format, not the input format? Ditto for Time and DateTime below.

**Review by ekoleda-codaio — CHANGES_REQUESTED** [review]

**Comment by ekoleda-codaio** [comment]
> Any idea why we get these warnings with `make typedoc`?
> 
> ```
> Warning: NoAuthentication, defined at types.ts:134, is referenced by Authentication but not included in the documentation.
> Warning: VariousAuthentication, defined at types.ts:433, is referenced by Authentication but not included in the documentation.
> Warning: HeaderBearerTokenAuthentication, defined at types.ts:240, is referenced by Authentication but not included in the documentation.
> Warning: CodaApiBearerTokenAuthentication, defined at types.ts:254, is referenced by Authentication but not included in the documentation.
> ```
> 
> I thought that typedoc pulled in the transitive closure of what was referenced in `index.ts`, but it appears to be omitting all these types that it warns about, for some reason. Some or all of those that it warns about we do want to be in the documentation.

I think TypeDoc only documents exported types. These warnings indicate that the signature of an exported type is referencing an unexported type.

File: helpers/url.ts (hunk 1 above)
Comment (jonathan-codaio) [reply helpers/url.ts]: Fixed.

File: helpers/url.ts (hunk 2 above)
Comment (jonathan-codaio) [reply helpers/url.ts]: Fixed.

File: schema.ts (hunk 3 above)
Comment (jonathan-codaio) [reply schema.ts]: Sure, updated all usage.

**Comment by jonathan-codaio** [comment]
> > Any idea why we get these warnings with `make typedoc`?
> > ```
> > Warning: NoAuthentication, defined at types.ts:134, is referenced by Authentication but not included in the documentation.
> > Warning: VariousAuthentication, defined at types.ts:433, is referenced by Authentication but not included in the documentation.
> > Warning: HeaderBearerTokenAuthentication, defined at types.ts:240, is referenced by Authentication but not included in the documentation.
> > Warning: CodaApiBearerTokenAuthentication, defined at types.ts:254, is referenced by Authentication but not included in the documentation.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I thought that typedoc pulled in the transitive closure of what was referenced in `index.ts`, but it appears to be omitting all these types that it warns about, for some reason. Some or all of those that it warns about we do want to be in the documentation.
> 
> I think TypeDoc only documents exported types. These warnings indicate that the signature of an exported type is referencing an unexported type.

Aha, interesting. Guess I'll start exporting these so I can document them.

**Review by ekoleda-codaio — APPROVED** [review]

**Pull request merged**